packages/api/src/controllers/stream.ts

```diff
@@ -1086,6 +1086,41 @@ app.put(
   }
 );
 
+app.post("/:id/lockPull", authorizer({ anyAdmin: true }), async (req, res) => {
+  const { id } = req.params;
+  let { leaseTimeout } = req.body;
+  if (!leaseTimeout) {
+    // Sets the default lock lease to 60s
+    leaseTimeout = 60 * 1000;
+  }
+  logger.info(`got /lockPull for stream=${id}`);
+
+  const stream = await db.stream.get(id, { useReplica: false });
+  if (!stream || (stream.deleted && !req.user.admin)) {
+    res.status(404);
+    return res.json({ errors: ["not found"] });
+  }
+  if (stream.isActive) {
+    return res.status(423);
+  }
+
+  const updateRes = await db.stream.update(
```

Comment:
WDYT of also adding a `WHERE` clause on this query instead, that disconsiders the stream if it has `isActive=false`? Given we don't have transactions on the DB access, this would guarantee we don't have race conditions here.

Reply:
I was curious about the race condition too, of two catalyst calls acquiring the lock at the same time. But is that what we're guarding against with this:
```
sql`COALESCE((data->'pull'->>'lockedAt')::bigint,0) < ${
        Date.now() - leaseTimeout
      }`,
```
Does that ensure the update doesn't go through if there's already an active lock? Or can there still be a race condition because they can happen in parallel with no transactions?

I'm not sure checking `isActive` would help that much because there is a fairly big gap between acquiring the lock and the stream becoming active? I may be misunderstanding your idea Victor

Reply:
I discussed it with Victor, it's fine as it is right now.

The JSON update is transactional, so it's ok.